From 8d47755837403a15f15874c1af303334feb48d9c Mon Sep 17 00:00:00 2001
From: Abishek Ganapathy <quic_abishekg@quicinc.com>
Date: Fri, 23 Aug 2024 16:03:26 +0530
Subject: [PATCH] cfg80211: Bypass link_id check for STA in nl80211_tx_mgmt

Link_id received from upper layer would not match with the link_id
present in wdev if the order of radios are different in AP and STA
resulting in EINVAL and returning.

To resolve this issue, bypass the link_id check if type is STA.

Change-Id: If43e1b35859514728665e056d6ec14f76faef094
Signed-off-by: Abishek Ganapathy <quic_abishekg@quicinc.com>
---
 net/wireless/nl80211.c | 11 +++++++----
 1 file changed, 7 insertions(+), 4 deletions(-)

diff --git a/net/wireless/nl80211.c b/net/wireless/nl80211.c
index 90e5c8a55172..7b64897485a1 100644
--- a/net/wireless/nl80211.c
+++ b/net/wireless/nl80211.c
@@ -12752,11 +12752,14 @@ static int nl80211_tx_mgmt(struct sk_buff *skb, struct genl_info *info)
 	 * the valid links for the _station_ anyway, so that's up
 	 * to the driver.
 	 */
-	if (params.link_id >= 0 &&
-	    !(wdev->valid_links & BIT(params.link_id))) {
-		wdev_unlock(wdev);
-		return -EINVAL;
+	if (wdev->iftype != NL80211_IFTYPE_STATION) {
+		if (params.link_id >= 0 &&
+		    !(wdev->valid_links & BIT(params.link_id))) {
+			wdev_unlock(wdev);
+			return -EINVAL;
+		}
 	}
+
 	wdev_unlock(wdev);
 
 	params.buf = nla_data(info->attrs[NL80211_ATTR_FRAME]);
-- 
2.34.1

